### ISSUE TITLE
Add tutorials

### ISSUE THREAD

**Issue by dpascualhe** [JdeRobot/PerceptionMetrics#245]
Add notebooks with simple tutorials. E.g., evaluating a UNet trained with pytorch against Rellis3D.

**Comment by Ravish990** [comment]
Hello @dpascualhe I am Ravish. I am studying at Polaris School Of Technology , where I am in first year . I want to contribute to the organization . I am very excited and ready to put my efforts in contributing to the organization . Can you please assign the issue to me to get started ?

**Comment by sergiopaniego** [comment]
Hi @Ravish990, thanks for your interest in contributing!

You can start by installing the package and working on the issue by suggesting your ideas on how to address it😄

**Comment by SakhinetiPraveena** [comment]
Hello @dpascualhe ,myself Praveena, currently working full time as a full stack application developer. I am starting out my open source journey. I want to contribute to this organisation for GSOC'25. Looking forward to learning and making significant contribution to the project.

I am trying to install and build the package following through Installation.md, I couldn't find the file CMakeList.txt hence cmake .. is giving me an error. I am using a mac machine. Could you help me navigate through the error?

**Comment by dpascualhe** [comment]
Hi @SakhinetiPraveena, thank you for your interest!

What docs are you using? Maybe you are referencing the docs for the previous DetectionMetrics version?

Our current codebase is purely python based and is distributed as a Python library + CLI, so there is no cmake involved. This info is available both in the README and [our website](https://jderobot.github.io/DetectionMetrics/). Relevant links:
- Installation instructions: https://jderobot.github.io/DetectionMetrics/v2/installation/
- Notebook tutorial: https://github.com/JdeRobot/DetectionMetrics/blob/master/examples/tutorial_image_segmentation.ipynb

**Comment by SakhinetiPraveena** [comment]
Thanks a lot for clarifying. I see that you want someone to review and give feedback on the tutorial. I'll get on it.

For your reference, this was the outdated one I was following https://jderobot.github.io/DetectionMetrics/v1/installation/

**Comment by SakhinetiPraveena** [comment]
Hi @dpascualhe,

I am using a mac machine and CUDA isn't supported on apple hardware. Can you suggest an alternative if it's feasible. If not I will shift to using a linux vm.

Also to give you an update I was able to successfully install the library. I was facing the issue with CUDA when I was trying out the example. Also if there is a requirement for any specific tutorial you want me to create, I can get started on it right away.

**Comment by dpascualhe** [comment]
> Hi [@dpascualhe](https://github.com/dpascualhe),
> 
> I am using a mac machine and CUDA isn't supported on apple hardware. Can you suggest an alternative if it's feasible. If not I will shift to using a linux vm.

Hi @SakhinetiPraveena , CUDA is recommended for acceleration, but it shouldn't be a hard requirement. Have you tried simply installing the latest pytorch version and running the notebook? It should work fine (slower tho). Otherwise, I believe pytorch supports GPU acceleration in MacOS although I don't know much about it (https://pytorch.org/docs/stable/notes/mps.html). 

> If there is a requirement for any specific tutorial you want me to create, I can get started on it or I can start with adding tests since no tests were added so far.
As of now, we only have a tutorial for image segmentation with pytorch. Potential tutorials to be developed are LiDAR segmentation, dataset conversion/merging, image segmentation with tensorflow... The scripts in the [examples directory](https://github.com/JdeRobot/DetectionMetrics/blob/master/examples) can give you some inspiration 😄

**Comment by SakhinetiPraveena** [comment]
Hi @dpascualhe , I tested out DetectionMetrics using image segmentation tutorial in my mac machine. I have added GPU acceleration for MacOS using MPS and raised the [Pull Request](https://github.com/JdeRobot/DetectionMetrics/pull/279 ) .

[torch.jit.load()](https://github.com/JdeRobot/DetectionMetrics/blob/cb7564faaf4f97f3456119c4eda5ceb0b0716e58/detectionmetrics/models/torch.py#L322) was also throwing an error while running on mac because of device mismatch. Adding the map_location argument  explicitly  resolves the issue and makes sure the model is loaded to the right device in all cases.

Do let me know if I need to make any modifications to this. Looking forward for your feedback!

**Comment by nithika987** [comment]
Hello @dpascualhe, this is Nithika Balaji, and I'm currently beginning my journey in open source and excited to contribute to the organization in GSoC 2025. 

I am facing some difficulties in running the example tutorial,(https://github.com/JdeRobot/DetectionMetrics/blob/master/examples/tutorial_image_segmentation.ipynb)

My error: "No valid deep learning framework found" when I try to import TorchImageSegmentationModel, even though I have done all necessary installations
- OS: Windows 11
- Torch version: 2.6.0+cpu
- CUDA available: False
- Open3D version: 0.19.0

Debugging attempts: 
1. I edited __init__.py in models to print the error and got the following:
- Torch not available No module named 'open3d._ml3d'
2. When I tried: python -c "import open3d.ml.torch as ml3d"
- error: Exception: Open3D was not built with PyTorch support!

Could you guide me on how to resolve this?
Thanks in advance!

**Comment by nithika987** [comment]
Hello @dpascualhe, this is Nithika Balaji, and I'm currently beginning my journey in open source and excited to contribute to the organization in GSoC 2025.

I am facing some difficulties in running the example tutorial,(https://github.com/JdeRobot/DetectionMetrics/blob/master/examples/tutorial_image_segmentation.ipynb)

My error: "No valid deep learning framework found" when I try to import TorchImageSegmentationModel, even though I have done all necessary installations
- OS: Windows 11
- Torch version: 2.2.2+cpu
- CUDA available: False
- Open3D version: 0.19.0

Debugging attempts:
1. I edited init.py in models to print the error and got the following:
- Torch not available No module named 'open3d._ml3d'
2. When I tried: python -c "import open3d.ml.torch as ml3d"
- error: Exception: Open3D was not built with PyTorch support!

Could you guide me on how to resolve this?
Thanks in advance!

**Comment by dpascualhe** [comment]
Hello @nithika987 ! Thanks for testing out our tutorial 😄 

Regarding `Open3D was not built with PyTorch support!`, it seems that Open3D-ML is not supported in Windows as of now: https://github.com/isl-org/Open3D/issues/6519.

However, if you are not working LiDAR data (which you aren't as the tutorials only covers image segmentation), you shouldn't have any issues. We have recently merged some changes to wrap up every Open3D-ML import clause and avoid crashing if it is not available: https://github.com/isl-org/Open3D/issues/6519

Could you debug and share the exact line that is yielding the exception reported?

Thanks!

**Comment by nithika987** [comment]
Hi @dpascualhe thanks for getting back!
**The steps I followed for the setup:**
mkdir .venv
python -m venv .venv
python.exe -m pip install --upgrade pip
pip install torch==2.2.2 torchvision==0.17.2
pip install -e .
(Additionally I had to connect to Jupyter  through:
pip install ipykernel
python -m ipykernel install --user --name=.venv --display-name "Python (.venv)"  
)
Versions installed: 
**Torch: 2.2.2+cpu | TorchVision: 0.17.2+cpu | Open3D: 0.19.0**

**The error:**
Torch not available No module named 'open3d._ml3d'
Tensorflow not available No module named 'tensorflow'
Exception: 
File ~\jderobot\DetectionMetrics\detectionmetrics\models\__init__.py:22
     19     print("Tensorflow not available",e)
     21 if not REGISTRY:
---> 22     raise Exception("No valid deep learning framework found")

Exception: No valid deep learning framework found

**Debugging:** 
1. I got the below even though pytorch installed before open3d, open3d not automatically detecting pytorch installation
python -c "import open3d; print(open3d._build_config)"

{'BUILD_TENSORFLOW_OPS': False, **'BUILD_PYTORCH_OPS': False**, 'BUILD_CUDA_MODULE': False, 'BUILD_SYCL_MODULE': False, 'BUILD_AZURE_KINECT': True, 'BUILD_LIBREALSENSE': True, 'BUILD_SHARED_LIBS': False, 'BUILD_GUI': True, 'ENABLE_HEADLESS_RENDERING': False, 'BUILD_JUPYTER_EXTENSION': True, 'BUNDLE_OPEN3D_ML': False, 'GLIBCXX_USE_CXX11_ABI': True, 'CMAKE_BUILD_TYPE': 'Release', 'CUDA_VERSION': '', 'CUDA_GENCODES': '', 'Tensorflow_VERSION': '', **'Pytorch_VERSION': '',** 'WITH_OPENMP': True}

2. I tried installing requirements using " https://github.com/isl-org/Open3D-ML/blob/main/requirements-torch.txt " from official repo, but still facing same error

Is there a specific way to install open3d ML to recognize pytorch? Or should I just switch to linux VM? Would appreciate any suggestions! Thanks!

**Comment by nithika987** [comment]
Hi @dpascualhe Nithika Balaji here again, I've identified the issue causing the error: _"Torch not available No module named 'open3d.ml3d'"_

**Issue 1:**  

- torch.py has => from detectionmetrics.models import torch_model_utils as tmu ;  import detectionmetrics.utils.lidar as ul
- Both import open3d._ml3d : which is not supported in Windows

**Solution: Creating a mock module:** mocks the open3d._ml3d module on Windows by injecting a custom MockModule into sys.modules and modifying sys.meta_path to ensure Python finds it. This allows code that imports open3d._ml3d to run without errors, even if the actual module is missing. 

Eg: open3d_ml3d_patch.py ->displays an error only when ml3d functions called in Windows (works as usual in Linux), doesn't raise error on just importing open3d._ml3d in Windows.
Must be imported into the __init__.py files.

Test 1: python -c "from detectionmetrics.models import TorchImageSegmentationModel; import torch; import open3d._ml3d; print(torch.__version__)"

Output: 2.2.2+cpu

**Issue 2: tutorial_image_segmentation.ipynb in examples**=> When system doesn't support CUDA

- Device compatibility in PyTorch, while calling TorchImageSegmentationModel(), model was saved with CUDA (GPU) and I'm  running it on CPU. The error occurs because aten::empty_strided is trying to use CUDA, but my system does not support it. 

- **Solution**: Converting the model to CPU before loading it.

**Comment by nithika987** [comment]
Hi @dpascualhe,

Regarding my fix for Issue 1 (mocking open3d._ml3d on Windows), does this approach seem feasible to you? If so, would it be appropriate for me to raise a PR for it?

Looking forward to your thoughts!

**Comment by dpascualhe** [comment]
Hi @nithika987 ! Could you please make sure to pull the latest commit from the master branch? I think you might be facing some issues that should be solved by now.

> **Issue 1:**
> 
> * torch.py has => from detectionmetrics.models import torch_model_utils as tmu ;  import detectionmetrics.utils.lidar as ul
> * Both import open3d._ml3d : which is not supported in Windows

In the latest version, all imports for Open3D-ML are wrapped in try/except clauses like these:

```
try:
    from open3d._ml3d.torch.models.kpconv import batch_grid_subsampling, batch_neighbors
except Exception:
    print("Open3D-ML3D not available")
```

You should see a trace saying 'Open3D-ML not available', but the exception should not be raised. Links to all relevant lines (https://github.com/JdeRobot/DetectionMetrics/blob/ef8942f3a229d3a553c7d14dfb2009537fbcf148/detectionmetrics/models/torch_model_utils/o3d_randlanet.py#L5, https://github.com/JdeRobot/DetectionMetrics/blob/ef8942f3a229d3a553c7d14dfb2009537fbcf148/detectionmetrics/models/torch_model_utils/o3d_randlanet.py#L5, https://github.com/JdeRobot/DetectionMetrics/blob/ef8942f3a229d3a553c7d14dfb2009537fbcf148/detectionmetrics/models/torch_model_utils/__init__.py#L5)

> **Issue 2: tutorial_image_segmentation.ipynb in examples**=> When system doesn't support CUDA
> 
> * Device compatibility in PyTorch, while calling TorchImageSegmentationModel(), model was saved with CUDA (GPU) and I'm  running it on CPU. The error occurs because aten::empty_strided is trying to use CUDA, but my system does not support it.

#279, which is already merged, should solve said issue, as we are now explicitly setting the `map_location` parameter to whatever device is available (preference order CUDA, MPS, cpu)

Please ensure that you are using the latest code and let me know if your issues are solved!

**Comment by nithika987** [comment]
Thank you so much for the clarification @dpascualhe and sorry for the trouble :(

**Comment by nithika987** [comment]
Hii @dpascualhe I'm not facing any issues now, thankss !!

**Comment by jayzalani** [comment]
Hey! I did check out the Databases that are supported but did not see [Rellis3d](https://github.com/unmannedlab/RELLIS-3D) can i try to add this along with its tutorial?

**Comment by dpascualhe** [comment]
> Hey! I did check out the Databases that are supported but did not see [Rellis3d](https://github.com/unmannedlab/RELLIS-3D) can i try to add this along with its tutorial? [@dpascualhe](https://github.com/dpascualhe)

Hi @jayzalani ! RELLIS-3D is supported ([Compatibility](https://jderobot.github.io/DetectionMetrics/v2/compatibility/)). In fact, the [current tutorial](https://github.com/JdeRobot/DetectionMetrics/blob/master/examples/tutorial_image_segmentation.ipynb) makes use of RELLIS-3D as the evaluation dataset. Feel free to give it a try and let us know how it went! 😄

**Comment by jayzalani** [comment]
ohh! thanks I was exploring [this](https://jderobot.github.io/DetectionMetrics/v1/resources/datasets/). That's great will explore this and try to build some good tutorial in This dataset. 😎

**Comment by dpascualhe** [comment]
Hi @jayzalani , sorry for the confusion, but I think you are looking into the docs for DetectionMetrics v1, which is no longer mantained. Your entrypoint should either be our [updated website](https://jderobot.github.io/DetectionMetrics/) or the [README](https://github.com/JdeRobot/DetectionMetrics/blob/master/README.md). As stated there:

> Now, we're excited to introduce DetectionMetrics v2! While retaining the flexibility of our previous release, DetectionMetrics has been redesigned with an expanded focus on image and LiDAR segmentation. As we move forward, **v2 will be the actively maintained version**, featuring continued updates and enhancements to keep pace with evolving AI and computer vision technologies.

Relevant links for the current version:

💻 [Code](https://github.com/JdeRobot/DetectionMetrics)	🔧 [Installation](https://jderobot.github.io/DetectionMetrics/v2/installation)	🧩 [Compatibility](https://jderobot.github.io/DetectionMetrics/v2/compatibility)	📖 [Docs](https://jderobot.github.io/DetectionMetrics/py_docs/_build/html/index.html)

> Would a tutorial demonstrating how to fine-tune pre-trained models on challenging Rellis3D samples be valuable for the project?

Fine-tuning or other training stages are not the main focus of DetectionMetrics. Our goal is to unify the evaluation stage for different frameworks and datasets. In that sense, new tutorials could focus in either using other datasets or frameworks (as of now our tutorial uses PyTorch and RELLIS-3D), or exploring other tools in the repo like computational cost estimation.

**Comment by tejasstanley** [comment]
Hi @dpascualhe,
Tejas Stanley here. I am currently starting my open source contribution journey and was thinking of creating a dataset conversion tutorial that adapts nuScenes to the DetectionMetrics format. Please let me know if this would be a suitable addition to the current tutorials.

**Comment by Khiladi-786** [comment]
Thanks for the clarification!

That makes sense — I understand that a RELLIS-3D dataset-based LiDAR segmentation tutorial would be more valuable.

I’ve already been working on a beginner-friendly version to explain the evaluation pipeline, and I can extend it to include:
- Proper RELLIS-3D dataset loading
- Real LiDAR samples + labels
- Evaluation using the existing PerceptionMetrics pipeline

I’ll align my work accordingly so it matches the expected dataset-based tutorial.

Please let me know if you have any specific preferences for structure or workflow.

**Comment by roPatel69** [comment]
Hi, I am interested in working on this issue.

I have started exploring the repository and I would like to contribute by creating a simple tutorial notebook for beginners.

I will first try to understand the project setup and then propose a tutorial. Please let me know if this approach works.

Thank you!

**Comment by dpascualhe** [comment]
Hi @roPatel69 , sure! Open up a new issue with the tutorial you plan to add before raising the PR so that we can align priorities

**Comment by vianmangal** [comment]
hello @dpascualhe , i saw the current tutorials and noticed that the original PyTorch + RELLIS-3D image-segmentation example is already covered.

 i'd like to contribute an end-to-end RELLIS-3D LiDAR segmentation tutorial notebook using the existing Rellis3DLiDARSegmentationDataset and TorchLiDARSegmentationModel. The notebook would cover:

 - Environment and dependency setup
 - RELLIS-3D data preparation and expected directory structure
 - Loading an approved pretrained model
 - Single-sample inference and point-cloud visualization
 - Evaluation on a dataset split
and more

would this still be a useful addition? If so, which supported model/backend and pretrained checkpoint would you prefer for the tutorial? i can open a focused issue for the notebook before beginning implementation. lmk

**Comment by dpascualhe** [comment]
Hi @vianmangal ! as of now I think we got all basics covered regarding tutorials. If something new comes up in that regard we'll let you know. Closing this issue for now.
